opengovsg/FormSG
# feat(sms-limiting): admin facing frontend

## Problem
This PR adds the admin facing frontend for the sms limiting feature. This feature is dependent on [this PR](https://github.com/opengovsg/FormSG/pull/2276).

The frontend in this PR can be split into 3 portions, which will be covered below: 
1. loading state 
2. success state
3. error state 

## Solution
The approach adopted tries to be as non-intrusive as possible and has minimal behavioural changes wrt old behaviour. 

**loading state**
Only the toggle is disabled during the loading state, when we are fetching data from the backend. This is to prevent the admin from toggling the state and showing the modal when there isn't enough data to determine the admin verificaiton state. 

**success state**
On fetching counts successfully from the backend, we toggle and set the verification state based on the counts + the form's message service id. 

Because `$scope.field` is also read by the verified button component, a design choice was made to set shared props on this. Alternatively, we could create a new property on `$scope` and share it through sharing the property using 2 way data binding but it'll serve much of the same purpose and there is very little benefit to be gained because of the shift to react.   

**error state**
Main concerns with the display during error state is 
1. how to handle the toggle
2. keeping sms toggling consistent 

a few different flows were considered: 
1. don't disable the toggle - this was rejected because we do not want to mislead users who have exceeded into thinking that they can toggle on (only to be rejected on save). 
2. Toaster vs displaying infobox - felt that a toaster offered better user experience compared to an infobox because the toaster is displayed in response to an action.

**Breaking Changes** 

- [x] Yes - this PR contains breaking changes (dependent)
    - This PR queries a new endpoint to obtain the counts for the admin, which will break on rollback of server
    - If the frontend change (this PR) is rollbacked, this will be ok as it just means that the new endpoint won't be queried

## Manual Tests
1. Create a form 
2. Create a mobile field for the form
3. Make the form public
4. Duplicate the form

- [ ] Attempt to toggle the mobile field. A modal should show up, displaying the sms counts for the admin of the form 
- [ ] Go to the other form and request OTP verification on the public user facing frontend. On the first form, disable otp verification then turn it on again. An updated sms counts should be shown 
- [ ] Transfer ownership of the first form to another user. Turn mobile otp verification off then on again. The counts should be for the new form admin. 

**The below tests require db write permissions on staging**
- [ ] Edit the form so that it has a message service id. Attempt to toggle otp verification. This should succeed without any modal displaying
- [ ] Insert >10k documents sent under the admin into the sms counts table (see the block below for the commands). Open the mobile field. The OTP verification should be disabled.  
```
db.getCollection('smscounts').insertMany([10,000 of fake sms count document])
```

**fake sms count document**
```
 {
          "smsType" : "VERIFICATION",
          "form" : <formId here>,
          "formAdmin" : {
              "email" : <your email here>,
              "userId" : ObjectId(<your id here>)
          },
          "msgSrvcSid" : <form's twilio sid here>,
          "logType" : "SUCCESS",
          "createdAt" : ISODate("2021-06-30T07:41:53.768Z"),
          "isOnboardedAccount" : false,
          "__v" : 0
        }
```

- [ ] Edit the form in db and add a own custom message service id. Go to that form and attempt to toggle sms verification on. This should succeed without any modal showing.

## Screen recordings
**Initial loading state**
![recording (4)](https://user-images.githubusercontent.com/44049504/124697725-e55a2a00-df19-11eb-9c44-ef4fcdcf60ca.gif)

**Admin verification under limit**
![success](https://user-images.githubusercontent.com/44049504/124076686-2a451300-da79-11eb-95a3-0acb94823fdb.gif)

**Admin over limit**
![over limit](https://user-images.githubusercontent.com/44049504/124076428-cf132080-da78-11eb-91a6-7a29e0abf0d1.gif)

**Form has message service id**
![has messageid](https://user-images.githubusercontent.com/44049504/124076642-1ef1e780-da79-11eb-91f5-37960337bca8.gif)

**Fetching sms counts errored**
![recording (1)](https://user-images.githubusercontent.com/44049504/124076876-66787380-da79-11eb-99dc-8d6fe51c80f3.gif)
